# <center>**TalentPitch Challenge: Exploratory Data Analysis (EDA)**</center>

## **Indice de Contenido**
<a href="">1. Data Load</a>

## **1. Data set load**

In [39]:
#Import librery

#Data manipulation
import pandas as pd
import numpy as np

#Data vizualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
warnings.filterwarnings('ignore') #evitar los molestos avisos.
#from matplotlib import rcParams


NameError: name 'warnings' is not defined

In [2]:
#CSV file location and Dataset Assignment to DataFrame
users = pd.read_csv('./data/users.csv')
users_raw = pd.read_csv('./data/users_raw.csv')

In [3]:
users.head()

user_id desired_state                      last_role  \
0      1.0    freelancer                            CTO   
1   9670.0  entrepreneur   Líder en gestión de procesos   
2  23825.0    employment     Líder de proyectos Masivos   
3   2245.0    employment             Lider de desarollo   
4  24328.0    freelancer  Diseñadora Gráfica Freelancer   

                                    dreamt_companies  \
0                         ['https://www.apple.com/']   
1  ['https://www.rutanmedellin.org/es/', 'http://...   
2  ['https://www.medtronic.com/co-es/index.html',...   
3                                 [None, None, None]   
4  ['https://www.kalley.com.co/', 'https://www.ma...   

                      forms_to_work  \
0  ['virtual', 'mixed', 'in_house']   
1             ['in_house', 'mixed']   
2  ['virtual', 'in_house', 'mixed']   
3                                []   
4                       ['virtual']   

                                     benefits_sought  \
0  ['bonuses', 'flexible_schedules', 'personal_re...   
1  ['bonuses', 'study_opportunities', 'scholarshi...   
2  ['flexible_schedules', 'study_opportunities', ...   
3                                                 []   
4  ['study_opportunities', 'flexible_schedules', ...   

                                     my_languages  \
0  [['ES', 'advanced'], ['EN', 'advanced', None]]   
1                                              []   
2                                              []   
3                            [[None, None, None]]   
4                            [[None, None, None]]   

                                last_study level_last_study  \
0                   Ingeniería de Sistemas            other   
1  Tecnología en sistemas electromecánicos            other   
2                            Bioingeniería              NaN   
3                   Ingeniería de Sistemas            other   
4               Gestión Del Diseño Gráfico            other   

                                    interested_areas  ...  \
0  ['Calidad de Software', 'Formación', 'Tecnolog...  ...   
1  ['Física', 'Diseño', 'Ecología', 'Tecnología',...  ...   
2  ['Starups', 'Marketing Digital Móvil', 'Progra...  ...   
3  ['Desarrollo de software', 'asf;aslkf;aslkfa;s...  ...   
4  ['Diseño gráfico', 'Mercadeo', 'marketing digi...  ...   

  views_to_resume_received views_to_profile_received saved_from_playlist  \
0                      NaN                       2.0                 NaN   
1                      NaN                       NaN                 NaN   
2                      NaN                       NaN                 NaN   
3                      NaN                       1.0                 NaN   
4                      NaN                       NaN                 NaN   

  reactions_received average_feedback  selector_appearances  reactiones_made  \
0                NaN             4.75                   NaN              NaN   
1                NaN              NaN                   NaN              NaN   
2                NaN              NaN                   NaN              NaN   
3                NaN              NaN                   NaN              NaN   
4                NaN              NaN                   1.0              NaN   

   match_with_playlists  contents_viewed  connections_sent  
0                   NaN              NaN               2.0  
1                   NaN              NaN               NaN  
2                   NaN              NaN               NaN  
3                   NaN              NaN               5.0  
4                   NaN              NaN               2.0  

[5 rows x 29 columns]

In [4]:
users_raw.head()

id           created_at   birthdate gender      city         state  \
0    1  2016-01-07 18:10:23  1978-11-13      M    Bogotá  Cundinamarca   
1   12  2016-01-07 18:10:23  1980-11-25      M  Medellín     Antioquia   
2  112  2016-01-18 17:28:56  1995-07-17      F  Envigado     Antioquia   
3  121  2016-01-18 17:28:57  1996-06-01      M  Medellín     Antioquia   
4  169  2016-01-18 17:29:03  1985-12-19      M  Medellin     Antioquia   

   profile_completed  
0               96.0  
1              100.0  
2               84.0  
3              100.0  
4               90.0

## **2. Data set exploration**

### **2.1 Users Dataset**

In [5]:
#General users Dataset information
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16189 entries, 0 to 16188
Data columns (total 29 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   user_id                    16189 non-null  float64
 1   desired_state              16189 non-null  object 
 2   last_role                  16168 non-null  object 
 3   dreamt_companies           16189 non-null  object 
 4   forms_to_work              16189 non-null  object 
 5   benefits_sought            16189 non-null  object 
 6   my_languages               16189 non-null  object 
 7   last_study                 16189 non-null  object 
 8   level_last_study           16072 non-null  object 
 9   interested_areas           16189 non-null  object 
 10  knowledges                 16189 non-null  object 
 11  skills                     16189 non-null  object 
 12  softwares_or_programs      16189 non-null  object 
 13  methodologies              16189 non-null  obj

In [6]:
#Checking the number of null values in the users dataset
users.isnull().sum()

user_id                          0
desired_state                    0
last_role                       21
dreamt_companies                 0
forms_to_work                    0
benefits_sought                  0
my_languages                     0
last_study                       0
level_last_study               117
interested_areas                 0
knowledges                       0
skills                           0
softwares_or_programs            0
methodologies                    0
roles                            0
received_messages            15940
sent_messages                16149
num_resumes_created          15812
events_scheduled             16153
views_to_resume_received     15921
views_to_profile_received    15446
saved_from_playlist          15894
reactions_received           16084
average_feedback             16112
selector_appearances         15207
reactiones_made              16122
match_with_playlists         16164
contents_viewed              16109
connections_sent    

In [7]:
#Distribution of numeric variables
users.describe()

user_id  received_messages  sent_messages  num_resumes_created  \
count  1.618900e+04         249.000000      40.000000           377.000000   
mean   1.576272e+05           1.224900       4.225000             3.649867   
std    3.032145e+05           0.974277      10.080414             8.091857   
min    1.000000e+00           1.000000       1.000000             1.000000   
25%    8.122400e+04           1.000000       1.000000             1.000000   
50%    9.954300e+04           1.000000       1.000000             2.000000   
75%    1.328010e+05           1.000000       2.000000             4.000000   
max    1.957403e+06          10.000000      52.000000           143.000000   

       events_scheduled  views_to_resume_received  views_to_profile_received  \
count         36.000000                268.000000                 743.000000   
mean          53.750000                  1.007463                   3.138627   
std          312.216992                  0.122169                   4.161614   
min            1.000000                  1.000000                   1.000000   
25%            1.000000                  1.000000                   1.000000   
50%            1.000000                  1.000000                   2.000000   
75%            2.000000                  1.000000                   3.000000   
max         1875.000000                  3.000000                  38.000000   

       saved_from_playlist  reactions_received  average_feedback  \
count           295.000000          105.000000         77.000000   
mean              2.230508            1.066667          3.923377   
std               2.008761            0.286446          1.002213   
min               1.000000            1.000000          1.000000   
25%               1.000000            1.000000          3.500000   
50%               1.000000            1.000000          4.110000   
75%               3.000000            1.000000          4.670000   
max              10.000000            3.000000          5.000000   

       selector_appearances  reactiones_made  match_with_playlists  \
count            982.000000        67.000000             25.000000   
mean               1.523422         3.626866              3.600000   
std                1.265459         6.694267              3.304038   
min                1.000000         1.000000              1.000000   
25%                1.000000         1.000000              1.000000   
50%                1.000000         1.000000              2.000000   
75%                2.000000         3.000000              6.000000   
max               10.000000        36.000000             11.000000   

       contents_viewed  connections_sent  
count        80.000000        461.000000  
mean          1.762500          2.869848  
std           2.177234         18.705277  
min           1.000000          1.000000  
25%           1.000000          1.000000  
50%           1.000000          2.000000  
75%           2.000000          2.000000  
max          19.000000        394.000000

In [8]:
#Distribution of categorical variables
users.describe(include=['O'])

desired_state   last_role dreamt_companies forms_to_work  \
count          16189       16168            16189         16189   
unique             5       10249             8417           105   
top       employment  Estudiante               []            []   
freq            9832         443             6621          4998   

       benefits_sought                    my_languages  \
count            16189                           16189   
unique            7348                             641   
top                 []  [['EN', 'intermediate', None]]   
freq              5422                            3621   

                    last_study level_last_study interested_areas knowledges  \
count                    16189            16072            16189      16189   
unique                    6947               11            11792      10508   
top     Ingeniería de Sistemas    undergraduate               []         []   
freq                       445             7195             3067       4932   

       skills softwares_or_programs methodologies  roles  
count   16189                 16189         16189  16189  
unique  11034                 10359          5795   2060  
top        []                    []            []     []  
freq     4976                  5177          6302  13628

### **2.1 Users Dataset**

In [11]:
#General user_raw Dataset information
users_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16189 entries, 0 to 16188
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 16189 non-null  int64  
 1   created_at         16188 non-null  object 
 2   birthdate          15783 non-null  object 
 3   gender             15548 non-null  object 
 4   city               15878 non-null  object 
 5   state              15852 non-null  object 
 6   profile_completed  16189 non-null  float64
dtypes: float64(1), int64(1), object(5)
memory usage: 885.5+ KB


In [10]:
#Checking the number of null values in the users_raw dataset
users_raw.isnull().sum()

id                     0
created_at             1
birthdate            406
gender               641
city                 311
state                337
profile_completed      0
dtype: int64

In [12]:
#Distribution of numeric variables
users_raw.describe()

id  profile_completed
count  1.618900e+04       16189.000000
mean   1.576272e+05          75.557292
std    3.032145e+05          24.364964
min    1.000000e+00          13.000000
25%    8.122400e+04          52.000000
50%    9.954300e+04          86.000000
75%    1.328010e+05          98.000000
max    1.957403e+06         100.000000

In [13]:
#Distribution of categorical variables
users_raw.describe(include=['O'])

created_at   birthdate gender    city         state
count                 16188       15783  15548   15878         15852
unique                16130        8179      4     601           137
top     2019-10-07 09:57:16  1998-11-03      M  Bogotá  Cundinamarca
freq                      6          10  10487    5285          6196

## **3. Union of dataset users and users_raw**

In [14]:
#Convirtiendo al mismo tipo de dato la columna id y user_id
users['user_id'] = users['user_id'].astype('int64')

#Visualizamos que la conversión haya sido exitosa
users['user_id'].drop_duplicates()
users.head()

user_id desired_state                      last_role  \
0        1    freelancer                            CTO   
1     9670  entrepreneur   Líder en gestión de procesos   
2    23825    employment     Líder de proyectos Masivos   
3     2245    employment             Lider de desarollo   
4    24328    freelancer  Diseñadora Gráfica Freelancer   

                                    dreamt_companies  \
0                         ['https://www.apple.com/']   
1  ['https://www.rutanmedellin.org/es/', 'http://...   
2  ['https://www.medtronic.com/co-es/index.html',...   
3                                 [None, None, None]   
4  ['https://www.kalley.com.co/', 'https://www.ma...   

                      forms_to_work  \
0  ['virtual', 'mixed', 'in_house']   
1             ['in_house', 'mixed']   
2  ['virtual', 'in_house', 'mixed']   
3                                []   
4                       ['virtual']   

                                     benefits_sought  \
0  ['bonuses', 'flexible_schedules', 'personal_re...   
1  ['bonuses', 'study_opportunities', 'scholarshi...   
2  ['flexible_schedules', 'study_opportunities', ...   
3                                                 []   
4  ['study_opportunities', 'flexible_schedules', ...   

                                     my_languages  \
0  [['ES', 'advanced'], ['EN', 'advanced', None]]   
1                                              []   
2                                              []   
3                            [[None, None, None]]   
4                            [[None, None, None]]   

                                last_study level_last_study  \
0                   Ingeniería de Sistemas            other   
1  Tecnología en sistemas electromecánicos            other   
2                            Bioingeniería              NaN   
3                   Ingeniería de Sistemas            other   
4               Gestión Del Diseño Gráfico            other   

                                    interested_areas  ...  \
0  ['Calidad de Software', 'Formación', 'Tecnolog...  ...   
1  ['Física', 'Diseño', 'Ecología', 'Tecnología',...  ...   
2  ['Starups', 'Marketing Digital Móvil', 'Progra...  ...   
3  ['Desarrollo de software', 'asf;aslkf;aslkfa;s...  ...   
4  ['Diseño gráfico', 'Mercadeo', 'marketing digi...  ...   

  views_to_resume_received views_to_profile_received saved_from_playlist  \
0                      NaN                       2.0                 NaN   
1                      NaN                       NaN                 NaN   
2                      NaN                       NaN                 NaN   
3                      NaN                       1.0                 NaN   
4                      NaN                       NaN                 NaN   

  reactions_received average_feedback  selector_appearances  reactiones_made  \
0                NaN             4.75                   NaN              NaN   
1                NaN              NaN                   NaN              NaN   
2                NaN              NaN                   NaN              NaN   
3                NaN              NaN                   NaN              NaN   
4                NaN              NaN                   1.0              NaN   

   match_with_playlists  contents_viewed  connections_sent  
0                   NaN              NaN               2.0  
1                   NaN              NaN               NaN  
2                   NaN              NaN               NaN  
3                   NaN              NaN               5.0  
4                   NaN              NaN               2.0  

[5 rows x 29 columns]

In [15]:
#Renombrando columna de id igual
users_raw.rename(columns={'id':'user_id'},
               inplace=True)
users_raw.columns

Index(['user_id', 'created_at', 'birthdate', 'gender', 'city', 'state',
       'profile_completed'],
      dtype='object')

In [16]:
#Uniendo dataset por id
users_join = pd.merge(users_raw,users, on='user_id',how='outer')
users_join

user_id           created_at   birthdate gender           city  \
0            1  2016-01-07 18:10:23  1978-11-13      M         Bogotá   
1           12  2016-01-07 18:10:23  1980-11-25      M       Medellín   
2          112  2016-01-18 17:28:56  1995-07-17      F       Envigado   
3          121  2016-01-18 17:28:57  1996-06-01      M       Medellín   
4          169  2016-01-18 17:29:03  1985-12-19      M       Medellin   
...        ...                  ...         ...    ...            ...   
16184  1957325  2023-04-14 16:19:33         NaN    NaN         Bogotá   
16185  1957347  2023-04-15 09:36:45  1997-08-05      F  Floridablanca   
16186  1957363  2023-04-15 16:03:38         NaN    NaN           Cali   
16187  1957391  2023-04-17 12:23:43         NaN    NaN           Cali   
16188  1957403  2023-04-17 16:53:22         NaN    NaN          Pasto   

                 state  profile_completed desired_state  \
0         Cundinamarca               96.0    freelancer   
1            Antioquia              100.0  entrepreneur   
2            Antioquia               84.0    employment   
3            Antioquia              100.0    employment   
4            Antioquia               90.0    employment   
...                ...                ...           ...   
16184     Cundinamarca               70.0    employment   
16185        Santander               52.0     first_job   
16186  Valle del Cauca               51.0    freelancer   
16187  Valle del Cauca               44.0    freelancer   
16188           Nariño               41.0    employment   

                                      last_role  \
0                                           CTO   
1                                           CEO   
2                       Auxiliar Administrativa   
3                                 Líder Talento   
4                       Coordinador de Sistemas   
...                                         ...   
16184                         Analista de Datos   
16185                        Diseñadora gráfica   
16186      Periodista, reportero e investigador   
16187                         Diseñador gráfico   
16188  Data analyst | Data Scientis Junior | BI   

                                        dreamt_companies  ...  \
0                             ['https://www.apple.com/']  ...   
1                                             ['Google']  ...   
2      ['https://gruponutresa.com/', 'https://www.seg...  ...   
3      ['https://cristianrestrepoq.wordpress.com', 'h...  ...   
4                        ['bancolombia', 'personalsoft']  ...   
...                                                  ...  ...   
16184                                ['Google, amazon,']  ...   
16185                             ['Google', 'facebook']  ...   
16186         ['BBC', 'The Guardian', 'Cambio Colombia']  ...   
16187              ['Organizaciones con trabajo remoto']  ...   
16188               ['Mercado Libre', 'Rappi', 'Google']  ...   

      views_to_resume_received views_to_profile_received saved_from_playlist  \
0                          NaN                       2.0                 NaN   
1                          NaN                       8.0                 NaN   
2                          NaN                       NaN                 NaN   
3                          NaN                      14.0                 1.0   
4                          NaN                       NaN                 NaN   
...                        ...                       ...                 ...   
16184                      NaN                       1.0                 NaN   
16185                      NaN                       NaN                 NaN   
16186                      NaN                       NaN                 NaN   
16187                      NaN                       NaN                 NaN   
16188                      NaN                       NaN                 NaN   

      reactions_received average_feedback selector_appearances  \


## **4. Understanding the data**

EL conjunto de datos puede hacer referencia a un portal de empleo, en el que una muestra representativa indicó información asociada a sus datos sociodemográficos incluyendo sus apiraciones laborales, conocimientos y destrezas. Tambien comprende información referente a su aspiración laboral y las interacciones realizadas por la muestra en el portal o plataforma y las realizadas por terceros al perfil creado por la muestra.

Para entender realmente el conjunto de datos, se hace necesario estudiar el significado y la relevancia de cada variable:

### 4.1 Operacionalización de variables

### Variable asociadas al seguimiento en el portal
<li><b>created_at:</b> Fecha y hora de creación de perfil</li>
<li><b>profile_completed:</b> Porcentaje de perfil completado</li>

#### Variables sociodemográfica
<li><b>user_id:</b> Identificador de cada usuario</li>
<li><b>birthdate:</b> Fecha de nacimiento</li>
<li><b>gender:</b> Género</li>
<li><b>city:</b> Ciudad</li>
<li><b>state:</b> Departamento</li>
<li><b>roles:</b> Rol desempeñado actualmente</li>

#### Variables asociadas a los conocimientos y destrezas 
<li><b>my_languages:</b> Idiomas que maneja</li>
<li><b>last_study:</b> último estudio realizado</li>
<li><b>level_last_study:</b> nivel de educación del último estudio realizado</li>
<li><b>knowledges:</b> Conocimientos que tiene</li>
<li><b>skills:</b> Habilidades que tiene</li>
<li><b>softwares_or_programs:</b> Programas o software que maneja</li>
<li><b>methodologies:</b> Metodologías de trabajo que maneja </li>

#### Variables asociadas a las aspiraciones laborales 
<li><b>desired_state:</b> Estado de empleo deseado</li>
<li><b>last_role:</b> último rol desempeñado</li>
<li><b>dreamt_companies:</b> compañias sonadas para trabajar</li>
<li><b>forms_to_work:</b> Formas de trabajo que más les gusta</li>
<li><b>benefits_sought:</b> Beneficios que busca la muestra tener mientras trabaja</li>
<li><b>interested_areas:</b> Áreas donde les interesa trabajar</li>

#### Variables asociadas a la interacción con el perfil creado por la muestra
<li><b>received_messages:</b> cantidad de mensajes recibidos</li>
<li><b>sent_messages:</b> cantidad de mensajes envíados</li>
<li><b>num_ resumes_ created:</b> cantidad de número de currículums creados</li>
<li><b>events_scheduled:</b> cantidad de eventos programados</li>
<li><b>views _ to_ resume_ received:</b> cantidad de vistas al currículums recibidas</li>
<li><b>views_to_profile_received:</b> cantidad de vistas al perfil recibidas</li>
<li><b>saved_from_playlist:</b> cantidad de listas de reproducción guardada </li>
<li><b>reactions_received:</b> cantidad de Reacciones recibidas</li>
<li><b>average_feedback:</b> Promedio de retroalimentación recibida</li>

### Variables asociadas a la interacción de la muestra con la plataforma
<li><b>selector_appearances:</b> selector de apariencias</li>
<li><b>reactiones_made:</b> Cantidad de reacciones realizadas</li>
<li><b>match_with_playlists:</b> Cantidad de match con lista de reproducción</li>
<li><b>contents_viewed:</b> Cantidad de contenidos vistos</li>
<li><b>connections_sent:</b> Conexiones enviadas </li>

### 4.1 Problema e hipótesis
Los problemas e hipótesis estudiados a partir de los datos anteriores pueden ser diversos:
<li> Relaciones entre variables sociodemográfica y aspiraciones laborales, pudiendose encontrar.</li>
<li> Seguimiento al porcentaje de perfil completado en el tiempo segun algunas caracteristicas sociodemográfica como el género o ultimo estudio realizado.</li>
<li>Atractivo del perfil creado por la muestra a partir de las variables asociadas a la interacción realizada por terceros.</li>
<li>Relación entre las variables de conocimiento y destrezas asociadas al atractivo del perfil creado o a las variables asociadas a las aspiraciones laborales.</li>
    


## **5. Creating new variables**

In [17]:
users_join['birthdate'].drop_duplicates()

0        1978-11-13
1        1980-11-25
2        1995-07-17
3        1996-06-01
4        1985-12-19
            ...    
16155    1990-07-18
16159    2000-06-09
16161    2002-07-21
16170    1994-10-21
16176    1988-04-05
Name: birthdate, Length: 8180, dtype: object

In [18]:
from datetime import datetime, timedelta
users_join['birthdate']=pd.to_datetime(users_join['birthdate'], format="")
users_join.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16189 entries, 0 to 16188
Data columns (total 35 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   user_id                    16189 non-null  int64         
 1   created_at                 16188 non-null  object        
 2   birthdate                  15783 non-null  datetime64[ns]
 3   gender                     15548 non-null  object        
 4   city                       15878 non-null  object        
 5   state                      15852 non-null  object        
 6   profile_completed          16189 non-null  float64       
 7   desired_state              16189 non-null  object        
 8   last_role                  16168 non-null  object        
 9   dreamt_companies           16189 non-null  object        
 10  forms_to_work              16189 non-null  object        
 11  benefits_sought            16189 non-null  object        
 12  my_l

In [19]:
#Cálculo columna edad
users_join['birthdate'].dt.year
ahora = pd.Timestamp('now')
users_join['Edad'] = (ahora - users_join['birthdate']).astype('<m8[Y]')

In [20]:
#Comprobación de Columna edad agregada
users_join.columns

Index(['user_id', 'created_at', 'birthdate', 'gender', 'city', 'state',
       'profile_completed', 'desired_state', 'last_role', 'dreamt_companies',
       'forms_to_work', 'benefits_sought', 'my_languages', 'last_study',
       'level_last_study', 'interested_areas', 'knowledges', 'skills',
       'softwares_or_programs', 'methodologies', 'roles', 'received_messages',
       'sent_messages', 'num_resumes_created', 'events_scheduled',
       'views_to_resume_received', 'views_to_profile_received',
       'saved_from_playlist', 'reactions_received', 'average_feedback',
       'selector_appearances', 'reactiones_made', 'match_with_playlists',
       'contents_viewed', 'connections_sent', 'Edad'],
      dtype='object')

In [21]:
users_join['Edad'].describe()

count    15783.000000
mean        31.337198
std          8.928948
min          0.000000
25%         25.000000
50%         29.000000
75%         36.000000
max        136.000000
Name: Edad, dtype: float64

In [23]:
users_join['Edad'].unique()

array([ 44.,  42.,  27.,  26.,  37.,  31.,  36.,   6.,  49.,  47.,  25.,
        29.,  38.,  32.,  30.,  24.,  48.,  60.,  28.,  57.,  50.,  35.,
        41.,  45.,  43.,  39.,  17.,  52.,  23.,  22.,  46.,  33.,  34.,
        40.,  66.,  nan,  21.,  20.,  55.,  68.,  53.,  54.,  70.,  18.,
        56.,  59.,  51.,  64.,  74.,   1.,  58.,  63.,  61.,  62.,  19.,
        72.,  67., 136.,   2.,  65.,  16., 135., 134.,  69.,  79.,   4.,
         3.,   5.,   0.])

La variable edad, tiene tres (3) valores atípicos 136, 135 y 134, valores Nan, y ceros(0). Pudiendo ser producto de que la vaiable fecha de nacimiento presentó casillas vacías o con fechas átipicas.

## **6. Dataset cleanup**

### **6.1 Reconocimiento de datos ruidosos**

In [ ]:
#Variables con datos cero en dataset
cols = users_join[users_join == 0].count(axis=0)
cols[cols > 0]

Edad    11
dtype: int64

In [24]:
#General users Dataset information
users_join.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16189 entries, 0 to 16188
Data columns (total 36 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   user_id                    16189 non-null  int64         
 1   created_at                 16188 non-null  object        
 2   birthdate                  15783 non-null  datetime64[ns]
 3   gender                     15548 non-null  object        
 4   city                       15878 non-null  object        
 5   state                      15852 non-null  object        
 6   profile_completed          16189 non-null  float64       
 7   desired_state              16189 non-null  object        
 8   last_role                  16168 non-null  object        
 9   dreamt_companies           16189 non-null  object        
 10  forms_to_work              16189 non-null  object        
 11  benefits_sought            16189 non-null  object        
 12  my_l

In [27]:
#Checking the number of null values in the users dataset
users_join.isnull().sum()

user_id                          0
created_at                       1
birthdate                      406
gender                         641
city                           311
state                          337
profile_completed                0
desired_state                    0
last_role                       21
dreamt_companies                 0
forms_to_work                    0
benefits_sought                  0
my_languages                     0
last_study                       0
level_last_study               117
interested_areas                 0
knowledges                       0
skills                           0
softwares_or_programs            0
methodologies                    0
roles                            0
received_messages            15940
sent_messages                16149
num_resumes_created          15812
events_scheduled             16153
views_to_resume_received     15921
views_to_profile_received    15446
saved_from_playlist          15894
reactions_received  

Recordando que hay un total de 16189 casos y 35 columnas, por lo cual hay variables con una cantidad muy chica de valores flatantes que han sido reconocidos como NaN y otras con una cantidad importante. En total, hay de veintiun (21) variables con "valores faltantes (NanN)", de las cuales catorce (14) variables referentes a la interacción con el portal (received_messages, sent_messages, num_resumes_created, events_scheduled, views_to_resume_received, views_to_profile_received, saved_from_playlist, reactions_received, average_feedback, selector_appearances, reactiones_made, match_with_playlists, contents_viewed,connections_sent) presentan un porcentaje alto de valores "perdidos". 

No obstante estos valores vacios, por la naturaleza de las variables hacen referencia a un cero, por lo que se procede a reemplazar las celdas vacías por el valor de cero.

In [28]:
#imputing with a constant

users_join_clean = users_join.fillna({'received_messages':0,'sent_messages':0, 'num_resumes_created':0, 'events_scheduled':0, 'views_to_resume_received':0, 'views_to_profile_received':0, 'saved_from_playlist':0, 'reactions_received':0, 'average_feedback':0, 'selector_appearances':0, 'reactiones_made':0, 'match_with_playlists':0, 'contents_viewed':0,'connections_sent':0})
users_join_clean.isnull().sum()

user_id                        0
created_at                     1
birthdate                    406
gender                       641
city                         311
state                        337
profile_completed              0
desired_state                  0
last_role                     21
dreamt_companies               0
forms_to_work                  0
benefits_sought                0
my_languages                   0
last_study                     0
level_last_study             117
interested_areas               0
knowledges                     0
skills                         0
softwares_or_programs          0
methodologies                  0
roles                          0
received_messages              0
sent_messages                  0
num_resumes_created            0
events_scheduled               0
views_to_resume_received       0
views_to_profile_received      0
saved_from_playlist            0
reactions_received             0
average_feedback               0
selector_a

Inicialmente se detectó once (11) casos en la Variable Edad con el valor en cero y variables con vectores vacíos que hacen referencia al perfil incompleto como puede observarse:

In [45]:
#Perfil completado con respecto a variables que presentan valores "perdidos"
#pd.crosstab (users_join_clean.profile_completed,users_join_clean.level_last_study)
#users_join_clean.groupby(['Edad']).count()['profile_completed']
#users_join_clean.groupby('last_role','level_last_study','state', 'city', 'gender', 'birthdate', 'created_at').mean()
# mirar como se distribuyen los nulos en edad
#uno=(users_join_clean[users_join_clean['Edad'].isna()].groupby(['last_role', 'city']).count()['profile_completed']) 
## Scatter plot:
#sns.set()
#cols = ['last_role','level_last_study','state', 'city', 'gender', 'birthdate', 'created_at','Edad']
#plt.hist(
#sns.pairplot(users_join_clean[cols], size = 2.5)
#plt.show();
    
#users_join_clean.plot.scatter(x='last_role', y='profile_completed')
pd.crosstab (users_join_clean.last_role, users_join_clean.profile_completed)


profile_completed                                   13.0   16.0   19.0   \
last_role                                                                 
"Asesor"                                                0      0      0   
-                                                       0      0      0   
--                                                      0      0      0   
---                                                     0      0      0   
.                                                       0      0      0   
...                                                   ...    ...    ...   
Área comercial Comertex: comercializadora de te...      0      0      0   
Áreas de mercadeo                                       0      0      0   
Último rol desempeñado como Profesional Univers...      0      0      0   
Último rol desempeñado, desarrollador Wordpress...      0      0      0   
último rol desempeñado: programador en framewor...      0      0      0   

profile_completed                                   21.0   22.0   23.0   \
last_role                                                                 
"Asesor"                                                0      0      0   
-                                                       0      0      0   
--                                                      0      0      0   
---                                                     0      0      0   
.                                                       0      0      0   
...                                                   ...    ...    ...   
Área comercial Comertex: comercializadora de te...      0      0      0   
Áreas de mercadeo                                       0      0      0   
Último rol desempeñado como Profesional Univers...      0      0      0   
Último rol desempeñado, desarrollador Wordpress...      0      0      0   
último rol desempeñado: programador en framewor...      0      0      0   

profile_completed                                   25.0   26.0   27.0   \
last_role                                                                 
"Asesor"                                                0      0      0   
-                                                       0      0      0   
--                                                      0      0      0   
---                                                     0      0      0   
.                                                       0      0      0   
...                                                   ...    ...    ...   
Área comercial Comertex: comercializadora de te...      0      0      0   
Áreas de mercadeo                                       0      0      0   
Último rol desempeñado como Profesional Univers...      0      0      0   
Último rol desempeñado, desarrollador Wordpress...      0      0      0   
último rol desempeñado: programador en framewor...      0      0      0   

profile_completed                                   29.0   ...  91.0   92.0   \
last_role                                                  ...                 
"Asesor"                                                0  ...      0      0   
-                                                       0  ...      0      1   
--                                                      0  ...      0      0   
---                                                     0  ...      0      0   
.                                                       0  ...      0      0   
...                                                   ...  ...    ...    ...   
Área comercial Comertex: comercializadora de te...      0  ...      0      0   
Áreas de mercadeo                                       0  ...      1      0   
Último rol desempeñado como Profesional Univers...      0  ...      0      0   
Último rol desempeñado, desarrollador Wordpress...      0  ...      0      0   
último rol desempeñado: programador en framewor...      0  ...      0      0   

profile_completed              

Sin embargo, la variable de edad podemos imputarla para no tener valores descriptivos con desviaciones altas.

In [ ]:
# calcular mediana de Age para imputar
train['Age'].median()
# imputar valor para rellenar nulos
train['Age'] = train['Age'].fillna(28.0)
train[['Survived', 'Sex', 'Age', 'Pclass']].info()

In [ ]:
import missingno as msno

In [ ]:
# how many total missing values do we have?
total_cells = np.product(melb_data.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
(total_missing/total_cells) * 100

## **7. Variable relationship**

### Caracterización Sociodemográfica de la muestra

Participaron un total de 16188 personas, de los cuales el 75.7% (n=78) eran del sexo femenino, el 22.3% (n=23) fueron del sexo masculino y el 1.9% (n=2) prefirieron describirse como fuera del binarismo de género; con una edad media de 31.34 años (DE=0.90). El 91.3% (n=94) de los participantes fueron solteros, el 43.7% de estrato socioeconómico dos, el 65% cursan el primer semestre y el 54.4% pertenecían a la facultad de ciencias jurídicas y sociales en la Universidad Simón Bolívar de Barranquilla.  En la Tabla 4 se presenta información detallada sobre las características demográficas de las personas.

In [ ]:
users_join['gender'].value_counts()

In [ ]:
users_join['Edad'].describe()

In [ ]:
np.sort

In [ ]:
pd.crosstab(users_join.gender , tips.smoker)

In [ ]:
users_join.max()

In [ ]:
# agrupar por compañia y calcular el promedio por cada una
users_join.groupby('city').mean()

## **6. Visualization**